src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,108 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        private static partial bool TryCloneFile(string sourceFullPath, in Interop.Sys.FileStatus srcStat, string destFullPath, bool overwrite)
+        {
+            // Try to clone the file immediately, this will only succeed if the destination doesn't exist,
+            // so we don't worry about locking for this one.
+            if (Interop.@libc.clonefile(sourceFullPath, destFullPath, Interop.@libc.CLONE_ACL) == 0)
+            {
+                // Success
+                return true;
+            }
+            Interop.Error error = Interop.Sys.GetLastError();
+
+            // Try to delete the destination file if we're overwriting.
+            if (error == Interop.Error.EEXIST && overwrite)
+            {
+                // Get the file permissions for source file
+                UnixFileMode filePermissions = (UnixFileMode)srcStat.Mode & SafeFileHandle.PermissionMask;
+
+                // Read FileStatus of destination file to determine how to continue (we need to check that
+                // destination doesn't point to the same file as the source file so we can fail appropriately)
+                int destError = Interop.Sys.Stat(destFullPath, out Interop.Sys.FileStatus destStat);
```

Comment:
Can you check the behavior of the `CopyOntoSelf` test?

Reply:
Nevermind, it's unneeded since we already have a lock on the source file, and we try to lock the destination file before deleting it. Have made these changes

Reply:
Yes, that is also what happens for the regulra (non-clonefile) copy.